File: PrimeNumbers/primes.cpp
#include <vector>
#include <iostream>
#include <array>

using namespace std;


class Primes
{
public:
    std::vector<int> _known;
    Primes()
    {
        _known.push_back(2);
        findupto(100);

    }
    void findupto(int upto)
    {
        for (int candidate = _known.front() + 1; candidate < upto; candidate++)
        {
            bool isPrime = true;
            for (auto itr = _known.begin(); itr != _known.end(); itr++)
            {
                int prime = *itr;
                int fraction = candidate / prime;
                if (fraction * prime == candidate)
                {
                    isPrime = false;
                }
            }
            if (isPrime)
            {
                _known.push_back(candidate);
            }
        }
    }

    bool isPrime(int x)
    {
        if (x > _known.back())
        {
            findupto(x);
        }
        
        return std::find(_known.begin(), _known.end(), x) != _known.end();
    }
    int& getElement(int index)
    {
        return _known.at(index);
    };
};

File: PrimeNumbers/primes.py


class Primes:
    def __init__(self):
        self._known = [2]
        self.find(limit=100)

    def find(self, limit):
        for candidate in range(self._known[-1], limit):
            is_prime = True
            for prime in self._known:
                fractional = int(candidate / prime)
                if fractional * prime == candidate:
                    is_prime = False
            if is_prime:
                self._known.append(candidate)

    def is_prime(self, x):
        if x > self._known[-1]:
            self.find(x)
        return x in self._known

    @property
    def known(self):
        return self._known


p = Primes()

print(17, 'isPrime', p.is_prime(17))
print(1500, 'isPrime', p.is_prime(1500))

print(p.known)